src/main/flight/pid_init.c

```diff
@@ -110,7 +110,7 @@ void pidInitFilters(const pidProfile_t *pidProfile)
     }
 #endif
 
-    if (dterm_lowpass_hz > 0 && dterm_lowpass_hz < pidFrequencyNyquist) {
```

Comment:
I don't see the point in changing this. While the PT1/2/3 filters may not experience the mathematical anomaly that BIQAUD does, what's the value in setting the cutoff above the nyquist? At best it will result in no filtering anyway so why even enable the filter and risk possible bad side effects?

Reply:
I really don't think this is resolved. What is the value in allowing the cutoff to be set above the nyquist for PT1/2/3? The nature of digital sampling is that there is no valid frequency signal above half the data sample rate (the nyquist frequency). And even so, what's the value in being able to set a cutoff over 4KHz for the gyro sampling at 8KHz?

Reply:
I agree with @etracer65 here - there do not seem to be any downsides to checking for situations where it is impossible for the filtering to have any possible beneficial effect, and avoiding the overhead of filtering in these situations.

Reply:
Ah, that's a different question entirely.
We agree that a check is required for biquads because they become unstable close to and above Nyquist.  No question about that.
PTn filters do not become unstable close to or above Nyquist since they only become unstable if k is greater than 1, and that cannot happen, regardless of cutoff.
I should emphasise that from a lowpass filter perspective, the Nyquist frequency has no specific relevance, provided that a filter is stable when set to frequencies above Nyquist, as is the case for our PTn type filters.  
To clarify what I mean, imagine a step input, as we get with RC command steps.  The height of the step after filtering is attenuated by the filter k value.  If this k value is 1.0, there is no filtering.  If it is 0.99999, there is barely any filtering.  Our PT code causes k to converge to 1.0 as the cutoff gets higher.  If, with a loop time of 4000hz, we were to set the cutoff to 8000hz, we would get k=0.926288318, which provides some, but not much, filtering.  More importantly, it causes no problem at all.
That's why there is clearly value in doing the check repeatedly every PID loop for the PT series filters.  No harm comes from setting the filter to even ridiculously high values.  
However @mikeller you raise the interesting question of whether we should limit the maximum upper cutoff value to some arbitrary number at which we consider there to be 'no possible beneficial effect'.
That requires consideration of the usage to which the filter is being put.
For dynamic notches, the user cannot currently set the maximum cutoff above 1000hz, as defined in `DYN_LPF_FILTER_FREQUENCY_MAX`.  That's a reasonable limit.  It is a value that has utility (eg provide square wave smoothing), and be perfectly safe, even if the FC is running at 1khz.  Hence for the dynamic notches, there is absolutely no need to do the pointless 'Nyquist' check every single loop, absolutely none.
For static notches, we (strangely) permit a maximum value for `FILTER_FREQUENCY_MAX`of 4000hz, with the note "maximum frequency for filter cutoffs (nyquist limit of 8K max sampling)".  
A 4000hz PT1 returns a k value of 0.75 for an 8k PID loop and 0.86 for a 4k PID loop.  That isn't a lot of change, packet to packet, but it is some.  
The delay from a 4000hz PT1 is 0.03ms, a trivial amount by any reckoning.  
I am not sure why the 4000hz limit was chosen for static filters, but as explained above, it does no harm, and it does some filtering, just very little.
If we leave this max as is, and a user does set it to max to see how that looks, it is probably not much more expensive in CPU cycles to do the filtering than to do the check.  However if we retain the check, we are doing that check for *everyone*, every single PID loop, even if they have the cutoff set to a reasonable value, just to help someone who sets it to a stupid value have some insignificant reduction in CPU usage (likely less than 0.001%).
So it seems pointless to penalise everyone with a totally unnecessary check of this kind just because it is possible, given our wide configuration range.
Maybe it may be simpler to set the `FILTER_FREQUENCY_MAX` value to 1000hz, so that if the user wants less filtering than that, they should just turn it off (which is what most people do anyway).
A 1000hz PT1 sets k to 0.61 at 4k and 0.43 at 8k.  Delay is 0.125ms, which some people prefer to minimise; they can just turn it off.  1000hz provides an adequate alternative to the two point averaging anti-aliasing filter when used in gyro lowpass2, whereas 4000hz would not have, so that's another reason to set the limit to 1000hz.
The only other situation where PT type filters may acquire very high cutoffs is in RC smoothing.  When the rc smoothing auto smoothing value is set to 0, the cutoffs are automatically set to 1.5x the RC link frequency.  Only for 1000khz links would this result in a value above 1000hz for the smoothing filter.  No such radios are currently available.  In any case, it doesn't matter.  So I don't see any need to do a 'Nyquist' check here either, given there is no safety issue, and that the intent is to smooth round corners off square steps, and that the filter will do this properly, safely and as intended for all possible RC link speeds.
I apologise for the long reply however this question really needs to be put to rest.
The check was, for PT filters, wrong in principle, and it totally unnecessary.
I will set `FILTER_FREQUENCY_MAX` to 1000hz so that users who want less filtering than that will be required to turn the filter off manually.  People may complain but lets see if they do.  I doubt it.

Reply:
I agree with @ctzsnooze. A check for nyquist limit is not needed for PTn filters as the calculated gain has a mathematical limit of 1.0f.

Above a gain of ~0.84 (at nyquist frequency) there doesn't exist a -3dB cutoff anymore, but there is still a usable cutoff that's weaker than -3dB up to a gain of 1.0. So PTn is always stable even above nyquist and we don't need a check.

**Btw. our BF gain calculation is just an approximation for small cutoff frequencies**, so at nyquist we actually haven't reached the real cutoff frequency at nyquist, yet. (blue vs. orange). Only at 0.8 * sampleFrequency (orange) is when our gain estimation hits the gain that corresponds to nyquist limit of 0.5 * sampleFrequency (blue).

![PT1_gain_overview](https://user-images.githubusercontent.com/19867640/118841091-49705280-b8c8-11eb-8755-ba518ffd3804.png)

Reply:
@ctzsnooze: Just to confirm my understanding of your response, your reasoning for not doing this check is that you want to save the effort of a single integer comparison on every single loop run by taking the risk of doing about 20 floating point operations which provide absolutely no value in every single loop run?
Or to ask the other way round, what do you see as the downside of leaving this sanity check in place?

Reply:
The check for both dynamic low passes, will be performed on all three axes for both gyro and D filters, or 6 times per PID loop, not just once.
Although I agree this is not much of a downside, 6 unnecessary operations are 6 unnecessary operations. 
The default settings involve cutoffs at values that do perform useful filtering.   So in the normal situation, these checks are entirely unnecessary.
In my extensive experience looking at logs, and I've looked at hundreds of logs a year from all kinds of users on Facebook and elsewhere, I have never, ever, seen a single log in which the user set one of these dynamic lowpass maximums above Nyquist.  And I've seen thousands of logs.  
In fact, I can't recall anyone setting a dynamic notch max above 1000hz - despite the maximum allowed being 4000hz.  I can't recall anyone advocating doing that.  People recognise, and are advised generally, that above about 1000hz very little is achieved, so they don't set the dynamic lowpass max that high.  Instead of setting the dynamic lowpass max above 1000hz, they turn the dynamic filter off.  That's the normal advice.  And once turned off, no unnecessary floating point operations are performed.  
Hence my proposal, already in this PR, which is to limit the configurable upper range of the dynamic lowpass filters to 1000hz.
Once that is done, the filter cannot be configured high enough by the user to *ever* trigger the Nyquist test, even if the test was left in place.  What point could there possibly be in doing a test 6 times a loop if we do that?  It will never, ever be positive.

Reply:
See gyro.h line 39:
```#define FILTER_FREQUENCY_MAX 1000 // so little filtering above 1000hz that if the user wants less delay, they must disable the filter``

Reply:
> The Nyquist check applies to both the gyro and D dynamic low passes by default, and will be performed on all three axes, or 6 times per PID loop - not just once per loop.
> Although I agree this is not much of a downside, 6 unnecessary operations are 6 unnecessary operations.

This is initialization code. It runs once. There is no performance justification for removing the nyquist check.

Reply:
Indeed.  Good point.  Apologies. Yes, this check only applies at init and that being so, I agree, there is no performance reason to remove the check.
But it remains unnecessary for the PTn filters.
And it would appear it is not going to protect the user very effectively with biquad filters.
It sees that the check on `dterm_lowpass_hz` confirms only, at init, that the *minimum* value of the dynamic lowpass is less than Nyquist.  
We never check the *maximum* value against Nyquist each loop as the cutoff value changes during the flight - therefore, if the user could set the maximum above Nyquist, we don't check for that, and there is nothing to prevent a dynamic cutoff going above Nyquist during the flight.
This would be a significant problem if a user set a biquad dynamic lowpass with, say, a minimum to 100hz, a maximum to 1000hz, and a loop time of 1000hz, for example.  The same values wouldn't be any problem with a PT1, PT2 or PT3, for reasons outlined above.
Fortunately, our existing CLI constraint on the highest possible minimum or maximum values of the dynamic notches is 1000hz, so we can't to get either cutoff close to Nyquist unless we run a 2k or less loop time, which is very unusual.  And who sets a minimum dynamic notch of 1000hz, other than unintentionally, 
It seems to me that if we are concerned about dynamic notches exceeding Nyquist, we should check that the *maximum* cannot exceed Nyquist at init time.  That's more important than the minimum, since it sets the highest value allowed.  This should definitely be done for the biquad.
However for the PT series there is no point worrying about the check.  The user can set the cutoff to wherever they like, nothing bad or unexpected will happen since 0<k<1 at all times.
Should we change the code to check max against Nyquist?  We never did that before.  If someone else wants to make a PR for that, they can.
In this PR however I want to put to rest the idea that doing the Nyquist check carries any benefit for people using PT filters, because it doesn't.  And since it provides no benefit, it should be removed.

Reply:
As you well know, once I found that biquad low passes have issues close to Nyquist, and ring with sharp square type waves, and knowing that they are unstable and would cause runaway oscillation if their cutoff exceeded Nyquist, this PR originally proposed that we dump the biquad.
Those who want it retained can make a different PR that solves the safety issues at low loop times and when cutoffs get close to Nyquist, but that's not something I feel comfortable doing.
I would be grateful if this PR can focus only on the addition of the PT2 and PT3 filter types and removing the unnecessary Nyquist checks for PT type filters. A separate PR can then be made to address the biquad issues that we have identified here.

Reply:
Having an explicit condition that allows people to easily understand that the filter will be **disabled** above the nyquist is better code then relying on people to implicitly understand that the filter is **effectively nonfunctional** above the nyquist. Restore the condition you removed. There is no "complicated" code to understand. Just put back the code you removed in this PR. Proposing that people make a follow-on PR to correct the mistakes in this one is not a way forward. I'm tired of discussing this and it's a waste of time.